app/Image/VideoHandler.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+namespace App\Image;
+
+use App\Exceptions\ConfigurationException;
+use App\Exceptions\ExternalComponentMissingException;
+use App\Exceptions\ImageProcessingException;
+use App\Exceptions\MediaFileOperationException;
+use App\Exceptions\MediaFileUnsupportedException;
+use App\Models\Configs;
+use FFMpeg\Coordinate\TimeCode;
+use FFMpeg\Exception\ExecutableNotFoundException;
+use FFMpeg\Exception\InvalidArgumentException;
+use FFMpeg\Exception\RuntimeException;
+use FFMpeg\FFMpeg;
+use FFMpeg\Media\Video;
+use Spatie\LaravelImageOptimizer\Facades\ImageOptimizer;
+
+class VideoHandler
+{
+	protected ?Video $video = null;
+
+	/**
+	 * "Loads" a video from the provided input file.
+	 *
+	 * **Warning:** Opposed to images, the video is not entirely loaded into
+	 * memory.
+	 * Hence, the physical on-disk representation of provided input file must
+	 * exist at least as long as the object of this class is used.
+	 * Otherwise, subsequent calls to method
+	 * (like {@link VideoHandler::saveFrame()}) which assume the file to exist
+	 * will fail.
+	 *
+	 * @throws ExternalComponentMissingException
+	 * @throws ConfigurationException
+	 * @throws MediaFileOperationException
+	 */
+	public function load(NativeLocalFile $file): void
+	{
+		if (!Configs::hasFFmpeg()) {
+			throw new ConfigurationException('FFmpeg is disabled by configuration');
+		}
+		try {
+			$ffmpeg = FFMpeg::create();
+			$audioOrVideo = $ffmpeg->open($file->getRealPath());
+			if ($audioOrVideo instanceof Video) {
+				$this->video = $audioOrVideo;
+			} else {
+				throw new MediaFileOperationException('No video streams found.');
+			}
+		} catch (ExecutableNotFoundException $e) {
+			throw new ExternalComponentMissingException('FFmpeg not found', $e);
+		} catch (InvalidArgumentException $e) {
+			throw new MediaFileOperationException('FFmpeg could not open media file', $e);
+		}
+	}
+
+	/**
+	 * Extracts and saves a frame from the video as image.
+	 *
+	 * @param NativeLocalFile $file          the file to write into
+	 * @param float           $framePosition the position of the frame to be extracted in seconds
+	 *
+	 * @return void
+	 *
+	 * @throws MediaFileOperationException
+	 */
+	public function saveFrame(NativeLocalFile $file, float $framePosition = 0.0): void
+	{
+		try {
+			$frame = $this->video->frame(TimeCode::fromSeconds($framePosition));
+			$frame->save($file->getRealPath());
+		} catch (RuntimeException $e) {
+			throw new MediaFileOperationException('Could not extract frame from video file', $e);
+		}
+		if (Configs::getValueAsBool('lossless_optimization')) {
+			ImageOptimizer::optimize($file->getRealPath());
+		}
```

Comment:
You are right. `extractFrame` is the only place where it is called and yes it is a waste of time. It is also an old "bug" which has been there a long time.

I see two options here

 1. Make the optimization upon save optional and let the caller decide whether optimization is needed at all, i.e. change the method signature to `saveFrame(bool $skipOptimization = false)`. The semantics would be as follows: If `$skipOptimzation == false` (the defautl), then `saveFrame` behaves as before, i.e. it optimizes the frame, if enabled by configuration. If `$skipOptimzation == true`, the optimization is always skipped regardless of the configuration. Calling function can use that, if they know that the will delete the result anyway.
 2. We can remove the call for optimization entirely. But if we decide to do it for video frames, we should do it consistently for the photo variants, too. As I pointed out in some other PHP comments, I am ensure how much the optimization helps at all. The docs can be found here: https://github.com/spatie/laravel-image-optimizer 
     - It is not consistently applied to the original size variant, because we copy that unaltered, if possible.
     - The optimization relies on external 3rd party binaries. If the binaries are not available, the optimization is a silent no-op. So it might be that users enable that option, but it actually does nothing and the user is not made aware of it.
     - The idea of laravel-image-optimizer is to optimize storage space my reducing the quality (see original docs) and removing file meta information such as EXIF data. Our configuration file resides is this [config/image-optimizer.php](https://github.com/LycheeOrg/Lychee/blob/master/config/image-optimizer.php). As far as I can tell, the external tools are indeed configured to be lossless. Hence, all what these tools seem to do is stripping of potential meta information. However, our size variants are stored by Imagick (or GD) and these libraries already remove the meta information. Hence, I wonder, if there is anything left what these tools could do. Maybe, the tools help to reduce storage space. For example, I don't know how good/bad Imagick is at storing WEBP images and whether the command line tool `cwebp` might be able to save some bytes even in lossless mode due to better compression algorithms. But I somehow doubt it, because the algorithms are mostly standardized as part of the format specification. And again, we are talking about the size variants here which are already reduced in resolution.

What option do you prefer?

Reply:
If practical, I would suggest merging `saveFrame` into `extractFrame` and removing the image optimization code.

For other uses of the image optimization code, I simply don't know how worthwhile it is. "Just leave it alone" would be my personal recommended approach; at least it defaults to off so it doesn't bother me or new users for that matter.

Reply:
Solutions can be so easy. :laughing: Fixed by https://github.com/LycheeOrg/Lychee/pull/1351/commits/62071f44f546f050395031dffa98eb76962b976c.